File: src/waypalm-site/src/main/java/waypalm/site/validation/UserEmailRegisteredValidator.java
package waypalm.site.validation;

import org.springframework.transaction.annotation.Transactional;
import waypalm.domain.dao.UserRepository;

import javax.inject.Inject;
import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class UserEmailRegisteredValidator implements ConstraintValidator<UserEmailRegistered, String> {
    private boolean shouldBeRegistered;

    @Inject
    protected UserRepository userRepository;

    @Override
    public void initialize(UserEmailRegistered constraintAnnotation) {
        initialize(constraintAnnotation.value());
    }

    public void initialize(boolean shouldBeRegistered) {
        this.shouldBeRegistered = shouldBeRegistered;
    }

    @Override
    @Transactional
    public boolean isValid(String value, ConstraintValidatorContext context) {
        return userRepository.findProfileByEmail(value) == null ^ shouldBeRegistered;
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/validation/UserEmailRegisteredValidator.kt
package waypalm.site.validation

import javax.validation.ConstraintValidator
import javax.inject.Inject
import waypalm.domain.dao.UserRepository
import javax.validation.ConstraintValidatorContext

public class UserEmailRegisteredValidator [Inject](
        val userRepository: UserRepository
)
: ConstraintValidator<UserEmailRegistered, String>
{
    var shouldBeRegistered: Boolean = false

    public override fun initialize(constraintAnnotation: UserEmailRegistered?) {
        shouldBeRegistered = constraintAnnotation!!.value()
    }

    public override fun isValid(value: String?, context: ConstraintValidatorContext?): Boolean {
        if(shouldBeRegistered)
            return userRepository.findProfileByEmail(value) != null
        else
            return userRepository.findProfileByEmail(value) == null
    }
}